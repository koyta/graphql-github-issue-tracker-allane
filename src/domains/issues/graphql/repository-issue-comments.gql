query RepositoryIssueComments(
  $repositoryOwner: String!
  $repositoryName: String!
  $issueNumber: Int!
  $first: Int!
  $after: String
) {
  repository(owner: $repositoryOwner, name: $repositoryName) {
    id
    issue(number: $issueNumber) {
      id
      comments(first: $first, after: $after) {
        pageInfo {
          hasNextPage
          endCursor
        }
        edges {
          cursor
          node {
            id
            body
            author {
              login
              avatarUrl
            }
          }
        }
      }
    }
  }
}
