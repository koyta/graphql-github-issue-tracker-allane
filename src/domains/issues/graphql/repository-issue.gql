query RepositoryIssue(
  $repositoryOwner: String!
  $repositoryName: String!
  $issueNumber: Int!
) {
  repository(owner: $repositoryOwner, name: $repositoryName) {
    id
    issue(number: $issueNumber) {
      id
      number
      title
      body
      author {
        login
      }
    }
  }
}
