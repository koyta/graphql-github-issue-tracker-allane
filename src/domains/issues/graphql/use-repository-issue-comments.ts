import { useCallback, useMemo } from "react";

import {
  RepositoryIssueCommentsQueryVariables,
  useRepositoryIssueCommentsQuery,
} from "@app/generated/graphql";

export const useRepositoryIssueComments = (
  variables: RepositoryIssueCommentsQueryVariables
) => {
  const { data, fetchMore, ...query } = useRepositoryIssueCommentsQuery({
    variables,
  });

  const { pageInfo, comments } = useMemo(
    () => ({
      pageInfo: data?.repository?.issue?.comments?.pageInfo,
      comments:
        data?.repository?.issue?.comments?.edges?.map((e) => e?.node) ?? [],
    }),
    [data]
  );

  const fetchMoreComments = useCallback(() => {
    if (!pageInfo?.hasNextPage) return;

    return fetchMore({
      variables: { after: pageInfo?.endCursor },
    });
  }, [fetchMore, pageInfo?.endCursor, pageInfo?.hasNextPage]);

  return { comments, fetchMoreComments, pageInfo, ...query };
};
