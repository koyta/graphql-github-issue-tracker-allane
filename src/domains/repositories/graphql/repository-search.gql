query RepositorySearch($query: String!, $first: Int, $after: String) {
  search(type: REPOSITORY, query: $query, first: $first, after: $after) {
    repositoryCount
    pageInfo {
      endCursor
      hasNextPage
    }
    edges {
      cursor
      node {
        ... on Repository {
          id
          name
          description
          url
          issues {
            totalCount
          }
          stargazers {
            totalCount
          }
          forks {
            totalCount
          }
          primaryLanguage {
            name
          }
          licenseInfo {
            name
          }
          owner {
            login
          }
        }
      }
    }
  }
}
