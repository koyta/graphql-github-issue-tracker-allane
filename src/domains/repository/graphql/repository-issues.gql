query RepositoryIssues(
  $repositoryName: String!
  $repositoryOwner: String!
  $states: [IssueState!]
  $first: Int!
  $after: String
  $filterBy: IssueFilters
) {
  repository(name: $repositoryName, owner: $repositoryOwner) {
    id
    issues(
      states: $states
      first: $first
      after: $after
      orderBy: { field: UPDATED_AT, direction: DESC }
      filterBy: $filterBy
    ) {
      totalCount
      pageInfo {
        hasNextPage
        endCursor
      }
      edges {
        cursor
        node {
          id
          number
          title
          body
          state
          comments {
            totalCount
          }
        }
      }
    }
  }
}
