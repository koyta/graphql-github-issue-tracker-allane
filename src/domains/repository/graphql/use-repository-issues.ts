import { useCallback, useMemo } from "react";

import {
  RepositoryIssuesQueryVariables,
  useRepositoryIssuesQuery,
} from "@app/generated/graphql";
import { notEmpty } from "@app/utils/notEmpty";

export const useRepositoryIssues = (
  variables: RepositoryIssuesQueryVariables
) => {
  const { data, fetchMore, ...query } = useRepositoryIssuesQuery({
    variables,
  });

  const { issues, pageInfo } = useMemo(
    () => ({
      issues: data?.repository?.issues.edges
        ?.map((edge) => edge?.node)
        .filter(notEmpty),
      pageInfo: data?.repository?.issues.pageInfo,
    }),
    [data]
  );

  const fetchMoreIssues = useCallback(() => {
    if (!pageInfo?.hasNextPage) {
      return;
    }

    return fetchMore({
      variables: { after: pageInfo?.endCursor },
    });
  }, [fetchMore, pageInfo, variables]);

  return {
    issues,
    pageInfo,
    fetchMoreIssues,
    ...query,
  };
};
