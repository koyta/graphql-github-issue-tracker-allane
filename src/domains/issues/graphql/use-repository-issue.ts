import { useRepositoryIssueQuery } from "@app/generated/graphql";

export const useRepositoryIssue = (variables: {
  repositoryName: string;
  repositoryOwner: string;
  issueNumber: number;
}) => {
  const { data, ...query } = useRepositoryIssueQuery({
    variables,
  });

  return {
    issue: data?.repository?.issue,
    ...query,
  };
};
