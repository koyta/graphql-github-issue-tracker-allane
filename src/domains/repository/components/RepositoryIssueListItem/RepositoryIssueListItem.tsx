import { Link } from "react-router-dom";

import { Card } from "@app/components/Card";
import { useRepositoryIssues } from "@app/domains/repository/graphql/use-repository-issues";
import { Issue, RepositoryIssuesQuery } from "@app/generated/graphql";

type Props = {
  issue: NonNullable<
    NonNullable<
      NonNullable<
        NonNullable<
          NonNullable<
            NonNullable<RepositoryIssuesQuery>["repository"]
          >["issues"]
        >["edges"]
      >[number]
    >["node"]
  >;
  issueUrl: string;
};

export const RepositoryIssueListItem = ({ issue, issueUrl }: Props) => {
  return (
    <Link key={issue.id} to={issueUrl} data-testid="issue-link">
      <Card>
        <h4 className="font-bold">{issue.title}</h4>
        <p className="text-base line-clamp-2">{issue.body}</p>
        <p>Comments: {issue.comments.totalCount}</p>
      </Card>
    </Link>
  );
};
