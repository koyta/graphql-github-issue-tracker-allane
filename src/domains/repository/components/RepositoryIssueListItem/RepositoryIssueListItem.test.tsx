import { ReactNode } from "react";

import { findByTestId, findByText, render } from "@testing-library/react";

import { IssueState } from "@app/generated/graphql";
import { RouterMock } from "@app/test/router-mock";

import { RepositoryIssueListItem } from "./RepositoryIssueListItem";

vi.mock("@app/components/Card", () => ({
  Card: ({ children }: { children: ReactNode }) => <div>{children}</div>,
}));

describe("RepositoryIssueListItem", () => {
  const issue = {
    id: "1",
    number: 100,
    body: "body",
    title: "title",
    state: IssueState.Open,
    comments: { totalCount: 10 },
  };
  const issueUrl = "/repos/owner/name/issues/100";

  it("should have tag <a> with the right href", async () => {
    const { container } = render(
      <RepositoryIssueListItem issue={issue} issueUrl={issueUrl} />,
      { wrapper: RouterMock }
    );
    const link = await findByTestId(container, "issue-link");
    expect(link).toHaveAttribute("href", issueUrl);
  });

  it("should display a title", async () => {
    const { container } = render(
      <RepositoryIssueListItem issue={issue} issueUrl={issueUrl} />,
      { wrapper: RouterMock }
    );
    expect(await findByText(container, issue.title)).toBeInTheDocument();
  });

  it("should display a body", async () => {
    const { container } = render(
      <RepositoryIssueListItem issue={issue} issueUrl={issueUrl} />,
      { wrapper: RouterMock }
    );
    expect(await findByText(container, issue.body)).toBeInTheDocument();
  });

  it("should display a comments count", async () => {
    const { container } = render(
      <RepositoryIssueListItem issue={issue} issueUrl={issueUrl} />,
      { wrapper: RouterMock }
    );
    expect(
      await findByText(container, `Comments: ${issue.comments.totalCount}`)
    ).toBeInTheDocument();
  });
});
