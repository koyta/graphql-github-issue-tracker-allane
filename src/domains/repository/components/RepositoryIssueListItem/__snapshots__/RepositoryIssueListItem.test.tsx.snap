// Vitest Snapshot v1

exports[`RepositoryIssueListItem > should render successfully 1`] = `
<body>
  <div>
    <a
      href="/repos/owner/name/issues/100"
    >
      <div
        class="flex flex-col p-2 rounded-md transition-colors hover:bg-gray-100 border-2 border-transparent hover:border-gray-100"
      >
        <h4
          class="font-bold"
        >
          title
        </h4>
        <p
          class="text-base line-clamp-2"
        >
          body
        </p>
        <p>
          Comments: 
          10
        </p>
      </div>
    </a>
  </div>
</body>
`;
